dotcom-rendering/src/web/components/Picture.tsx

```diff
@@ -140,21 +173,71 @@ export const Picture = ({
 	isMainMedia = false,
 	isLazy = true,
 }: Props) => {
-	const hdpiSources = getSources(role, 'hdpi', imageSources);
-	const mdpiSources = getSources(role, 'mdpi', imageSources);
-	const fallbackSrc = getFallback(role, 'hdpi', imageSources);
-	const sizes = getSizes(role, isMainMedia, format);
+	const hdpiSourceSets = getSourcesForRoleAndResolution(
+		imageSources,
+		role,
+		'hdpi',
+	);
+	const mdpiSourceSets = getSourcesForRoleAndResolution(
+		imageSources,
+		role,
+		'mdpi',
+	);
+	const fallbackSrc = getFallback(
+		hdpiSourceSets.length ? hdpiSourceSets : mdpiSourceSets,
+	);
+
+	// Create an array of breakpoints going from highest to lowest, with 0 as the final option
+	const breakpointSizes: [number, number][] = [
+		...Object.values(breakpoints).reverse(),
+		0,
+	].map((breakpoint) => [
+		breakpoint,
+		getSizeForBreakpoint(breakpoint, role, isMainMedia, format),
+	]);
+	const optimisedBreakPointSizes: [number, number][] =
+		optimiseBreakpointSizes(breakpointSizes);
 
 	return (
 		<picture itemProp="contentUrl">
-			{/* HDPI Source (DPR2) - images in this srcset have `dpr=2&quality=45` in the url */}
-			<source
-				srcSet={hdpiSources}
-				sizes={sizes}
-				media="(-webkit-min-device-pixel-ratio: 1.25), (min-resolution: 120dpi)"
-			/>
-			{/* MDPI Source (DPR1) - images in this srcset have `quality=85` in the url */}
-			<source srcSet={mdpiSources} sizes={sizes} />
+			{format.display === ArticleDisplay.Immersive && isMainMedia ? (
+				<>
+					{portraitImmersiveSource(hdpiSourceSets, true)}
+					{portraitImmersiveSource(mdpiSourceSets, false)}
+				</>
+			) : (
+				''
+			)}
+
+			{optimisedBreakPointSizes.map(([breakpoint, size]) => (
+				<>
+					{/* HDPI Source (DPR2) - images in this srcset have `dpr=2&quality=45` in the url */}
+					{hdpiSourceSets.length ? (
+						<source
+							srcSet={getSourceSetForSize(
+								size,
+								hdpiSourceSets,
+								true,
+							)}
+							sizes={`${breakpoint || size}px`}
+							media={`(min-width: ${breakpoint}px) and (-webkit-min-device-pixel-ratio: 1.25), (min-width: ${breakpoint}px) and (min-resolution: 120dpi)`}
+						/>
+					) : (
+						''
```

Comment:
`''`? Not sure if this is an issue or not but normally in this kind of situation you see `<></>`?

Reply:
Ah I'll definitely use that instead, I was looking for a similar situation to figure out what the best value to provide there was!